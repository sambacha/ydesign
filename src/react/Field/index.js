import React from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components';
import Box from '../Box';
import Text from '../Text';

const OptionalLabel = styled(Text)`
  &:after {
    content: ' (optional)';
    font-weight: normal;
    font-style: italic;
  }
`;

const Field = React.forwardRef(({ label, children, ...props }, ref) => {
  const isRequired = children => {
    let value = false;
    React.Children.forEach(children, child => {
      if (child.props.required) {
        value = true;
      }
    });
    return value;
  };

  const renderLabel = required => {
    if (required) {
      return (
        <Text fontSize={1} fontWeight={3} mb={2}>
          {label}
        </Text>
      );
    } else {
      return (
        <OptionalLabel fontSize={1} fontWeight={3} mb={2}>
          {label}
        </OptionalLabel>
      );
    }
  };

  return (
    <Box
      as={'label'}
      display={'inline-flex'}
      flexDirection={'column'}
      alignItems={'flex-start'}
      mb={3}
      ref={ref}
      {...props}
    >
      {renderLabel(isRequired(children))}
      {children}
    </Box>
  );
});

Field.propTypes = {
  // ...Box.propTypes,
  label: PropTypes.string.isRequired,
  children: PropTypes.element.isRequired,
};

Field.displayName = 'Field';

export default Field;
