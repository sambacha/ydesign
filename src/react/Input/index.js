import React from 'react'
import PropTypes from 'prop-types'
import styled from 'styled-components'
import { themeGet } from '@styled-system/theme-get'
import { CheckCircle } from '@yfi/icons/es/md'
import { Warning } from '@yfi/icons/es/md'
import FileInput from './FileInput'
import { StyledBox } from '../Box'

const StyledInput = styled(StyledBox).attrs(props => ({
  as: 'input',
}))`
  & {
    appearance: none;
  }

  &:hover {
    box-shadow: 0px 2px 6px rgba(0, 0, 0, 0.3);
  }

  &:focus {
    outline: none;
    border-color: ${themeGet('colors.primary', '#000')};
  }

  &:disabled {
    opacity: 0.4;
    pointer-events: none;
  }

  &[type="color"] {
    min-width: 3rem;
    padding: 0;
    &::-webkit-color-swatch-wrapper {
      display: flex;
      align-items: center;
      justify-content: center;
      padding: 0;
    }
    &::-webkit-color-swatch {
      flex: 0 0 auto;
      height: 2rem;
      width: 2rem;
      border: none;
      border-radius: 100%;
    }
    &::-moz-color-swatch {
      height: 2rem;
      width: 2rem;
      border: none;
      border-radius: 100%;
    }
  }

  .was-validated & {
    &:valid {
      border-color: ${themeGet('colors.success')};
    }
    &:valid ~ svg.icon-valid {
      visibility: visible;
      color: ${themeGet('colors.success')};
    }
    &:invalid {
      border-color: ${themeGet('colors.danger')};
    }
    &:invalid ~ svg.icon-invalid {
      visibility: visible;
      color: ${themeGet('colors.danger')};
    }
  }}
`

const StyledIconWrapper = styled.div`
  & {
    position: relative;
    display: flex;
    align-items: center;
    width: 100%;
  }
  > ${StyledInput} {
    padding-right: 3rem;
    text-overflow: ellipsis;
    white-space: no-wrap;
  }
  > svg {
    position: absolute;
    right: 1rem;
    visibility: hidden;
  }
`

const WithValidationStyle = React.forwardRef((props, ref) => {
  return (
    <StyledIconWrapper>
      <StyledInput {...props} ref={ref} />
      <CheckCircle className={'icon-valid'} />
      <Warning className={'icon-invalid'} />
    </StyledIconWrapper>
  )
})

const Input = React.forwardRef((props, ref) => {
  if (props.type === 'file') {
    return <FileInput {...props} ref={ref} />
  } else {
    return <StyledInput {...props} ref={ref} />
  }
})

const defaultProps = {
  color: 'text',
  bg: 'background',
  fontSize: '1rem',
  lineHeight: 'initial',
  height: '3rem',
  p: 3,
  border: 1,
  borderColor: 'moon-gray',
  borderRadius: 1,
  boxShadow: 1,
}

Input.WithValidationStyle = WithValidationStyle

WithValidationStyle.defaultProps = defaultProps
StyledInput.defaultProps = defaultProps
Input.defaultProps = defaultProps

Input.propTypes = {
  theme: PropTypes.object,
}

Input.displayName = 'Input'

export { StyledInput }

export default Input
