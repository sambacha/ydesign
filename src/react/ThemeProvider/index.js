import React from 'react';
import PropTypes from 'prop-types';
import { ThemeProvider as StyledThemeProvider } from 'styled-components';
import theme from '../theme';

const ThemeProvider = ({ children, theme, ...rest }) => {
  return (
    <StyledThemeProvider theme={theme} {...rest}>
      {children}
    </StyledThemeProvider>
  );
};

ThemeProvider.propTypes = {
  theme: PropTypes.object,
};

ThemeProvider.defaultProps = {
  theme,
};

ThemeProvider.displayName = 'ThemeProvider';

export default ThemeProvider;
