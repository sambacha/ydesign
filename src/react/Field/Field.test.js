import Field from './';
import Input from '../Input';
import ThemeProvider from '../ThemeProvider';
import theme from '../theme';

describe('Field component sanity', () => {
  it('has name', () => {
    expect(Field.displayName).toBe('Field');
  });

  it('matches default snapshot', () => {
    const component = render(
      <Field label="Input label">
        <Input />
      </Field>
    );
    expect(component).toMatchSnapshot();
  });

  it('matches themed snapshot', () => {
    const component = render(
      <ThemeProvider theme={theme}>
        <Field label="Themed input label">
          <Input />
        </Field>
      </ThemeProvider>
    );
    expect(component).toMatchSnapshot();
  });
});
