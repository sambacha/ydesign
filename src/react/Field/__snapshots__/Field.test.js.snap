// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Field component sanity matches default snapshot 1`] = `
.c0 {
  box-sizing: border-box;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-align-items: flex-start;
  -webkit-box-align: flex-start;
  -ms-flex-align: flex-start;
  align-items: flex-start;
  margin-bottom: 16px;
}

.c1 {
  box-sizing: border-box;
  font-size: 14px;
  font-weight: 600;
  margin-bottom: 8px;
  line-height: 1.5;
}

.c2:after {
  content: ' (optional)';
  font-weight: normal;
  font-style: italic;
}

.c3 {
  box-sizing: border-box;
  color: #3F3D4B;
  background-color: #fff;
  font-size: 1rem;
  line-height: initial;
  height: 3rem;
  padding: 16px;
  border: 1px solid transparent;
  border-color: #ccc;
  border-radius: 4px;
  box-shadow: 0px 2px 4px rgba(0,0,0,0.1);
}

.c3 {
  -webkit-appearance: none;
  -moz-appearance: none;
  appearance: none;
}

.c3:hover {
  box-shadow: 0px 2px 6px rgba(0,0,0,0.3);
}

.c3:focus {
  outline: none;
  border-color: #4E3FCE;
}

.c3:disabled {
  opacity: 0.4;
  pointer-events: none;
}

.c3[type="color"] {
  min-width: 3rem;
  padding: 0;
}

.c3[type="color"]::-webkit-color-swatch-wrapper {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  padding: 0;
}

.c3[type="color"]::-webkit-color-swatch {
  -webkit-flex: 0 0 auto;
  -ms-flex: 0 0 auto;
  flex: 0 0 auto;
  height: 2rem;
  width: 2rem;
  border: none;
  border-radius: 100%;
}

.c3[type="color"]::-moz-color-swatch {
  height: 2rem;
  width: 2rem;
  border: none;
  border-radius: 100%;
}

.was-validated .c3:valid {
  border-color: #28C081;
}

.was-validated .c3:valid ~ svg.icon-valid {
  visibility: visible;
  color: #28C081;
}

.was-validated .c3:invalid {
  border-color: #DC2C10;
}

.was-validated .c3:invalid ~ svg.icon-invalid {
  visibility: visible;
  color: #DC2C10;
}

<label
  class="c0"
  display="inline-flex"
>
  <div
    class="c1 c2"
    font-size="1"
    font-weight="3"
  >
    Input label
  </div>
  <input
    class="c3"
    color="text"
    font-size="1rem"
    height="3rem"
  />
</label>
`;

exports[`Field component sanity matches themed snapshot 1`] = `
.c0 {
  box-sizing: border-box;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-align-items: flex-start;
  -webkit-box-align: flex-start;
  -ms-flex-align: flex-start;
  align-items: flex-start;
  margin-bottom: 16px;
}

.c1 {
  box-sizing: border-box;
  font-size: 14px;
  font-weight: 600;
  margin-bottom: 8px;
  line-height: 1.5;
}

.c2:after {
  content: ' (optional)';
  font-weight: normal;
  font-style: italic;
}

.c3 {
  box-sizing: border-box;
  color: #3F3D4B;
  background-color: #fff;
  font-size: 1rem;
  line-height: initial;
  height: 3rem;
  padding: 16px;
  border: 1px solid transparent;
  border-color: #ccc;
  border-radius: 4px;
  box-shadow: 0px 2px 4px rgba(0,0,0,0.1);
}

.c3 {
  -webkit-appearance: none;
  -moz-appearance: none;
  appearance: none;
}

.c3:hover {
  box-shadow: 0px 2px 6px rgba(0,0,0,0.3);
}

.c3:focus {
  outline: none;
  border-color: #4E3FCE;
}

.c3:disabled {
  opacity: 0.4;
  pointer-events: none;
}

.c3[type="color"] {
  min-width: 3rem;
  padding: 0;
}

.c3[type="color"]::-webkit-color-swatch-wrapper {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  padding: 0;
}

.c3[type="color"]::-webkit-color-swatch {
  -webkit-flex: 0 0 auto;
  -ms-flex: 0 0 auto;
  flex: 0 0 auto;
  height: 2rem;
  width: 2rem;
  border: none;
  border-radius: 100%;
}

.c3[type="color"]::-moz-color-swatch {
  height: 2rem;
  width: 2rem;
  border: none;
  border-radius: 100%;
}

.was-validated .c3:valid {
  border-color: #28C081;
}

.was-validated .c3:valid ~ svg.icon-valid {
  visibility: visible;
  color: #28C081;
}

.was-validated .c3:invalid {
  border-color: #DC2C10;
}

.was-validated .c3:invalid ~ svg.icon-invalid {
  visibility: visible;
  color: #DC2C10;
}

<label
  class="c0"
  display="inline-flex"
>
  <div
    class="c1 c2"
    font-size="1"
    font-weight="3"
  >
    Themed input label
  </div>
  <input
    class="c3"
    color="text"
    font-size="1rem"
    height="3rem"
  />
</label>
`;
